org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend

```diff
@@ -411,11 +433,33 @@ import org.eclipse.xtext.validation.Issue
 	}
 
 	override codeLens(CodeLensParams params) {
-		throw new UnsupportedOperationException("TODO: auto-generated method stub")
+		return requestManager.runRead[ cancelIndicator |
+			val uri = params.textDocument.uri.toUri
+			val resourceServiceProvider = uri.resourceServiceProvider
+			val codeLensService = resourceServiceProvider?.get(ICodeLensService)
+			if (codeLensService === null)
+				return emptyList
+
+			return workspaceManager.doRead(uri) [ document, resource |
+				codeLensService.computeCodeLenses(document, resource, params, cancelIndicator)
+			]
+		]
 	}
 
 	override resolveCodeLens(CodeLens unresolved) {
-		return CompletableFuture.completedFuture(unresolved)
+		return CompletableFutures.computeAsync[ cancelChecker |
```

Comment:
Is not `doRead` here more appropriate, otherwise, this future won't be canceled by a write request.